### PR TITLE
Use coconut to support private coins in fastpay

### PR DESCRIPTION
**Pull request by ma2bd**
Use the coconut library (including bulletproofs) to implement fully "anonymous coins" (next to "transparent coins") in FastPay.

* Both coins are stored by their owner as a new data structure called `Asset`
* Includes two changes in Coconut: minor renaming + remove incorrect verification check in proofs
* Still lacking unit tests (in particular poking around all the security invariants)

### CODE REVIEW

File: fastpay/src/server.rs [hunk 1]
```diff
@@ -204,27 +278,76 @@ fn main() {
...
         }
+
+        ServerCommands::GenerateAll {
+            authorities,
+            committee,
+        } => {
+            let mut rng = coconut::rand::thread_rng();
+            let parameters = coconut::Parameters::new(
+                3,
+                /* TODO: check party capacity for bulletproofs */ authorities.len(),
```
Comment (ma2bd) [inline fastpay/src/server.rs]: @asonnino Is this correct?

File: coconut/src/proof.rs [hunk 2]
```diff
@@ -209,7 +209,6 @@ impl RequestCoinsProof {
         offset: &Scalar,
     ) -> CoconutResult<()> {
         ensure!(sigmas.len() == kappas.len(), CoconutError::MalformedProof);
-        ensure!(sigmas.len() == cms.len(), CoconutError::MalformedProof);
```
Comment (ma2bd) [inline coconut/src/proof.rs:212]: @asonnino I think this was a typo

File: fastpay/src/server.rs (hunk 1 above)
Comment (asonnino) [reply fastpay/src/server.rs]: No, this is the max number of output coins that can be produced. We use the aggregation feature of bulletproofs to pay log cost in the number of range proofs.

File: coconut/src/proof.rs (hunk 2 above)
Comment (asonnino) [reply coconut/src/proof.rs:212]: yes I think it is

**Review by asonnino — APPROVED** [review]

**Pull request merged**